test/extensions/filters/listener/tls_inspector/tls_inspector_test.cc

```diff
@@ -485,6 +487,407 @@ TEST_P(TlsInspectorTest, RequestedMaxReadSizeDoesNotGoBeyondMaxSize) {
   EXPECT_FALSE(io_handle_->isOpen());
 }
 
+void TlsInspectorTest::testJA4(const std::string& expected_ja4, const std::string& sni,
+                               const std::string& alpn) {
+  envoy::extensions::filters::listener::tls_inspector::v3::TlsInspector proto_config;
+  proto_config.mutable_enable_ja4_fingerprinting()->set_value(true);
+  cfg_ = std::make_shared<Config>(*store_.rootScope(), proto_config);
+
+  std::vector<uint8_t> client_hello =
+      Tls::Test::generateClientHello(std::get<0>(GetParam()), std::get<1>(GetParam()), sni, alpn);
+
+  init();
+  mockSysCallForPeek(client_hello);
+
+  EXPECT_CALL(socket_, setJA4Hash(absl::string_view(expected_ja4)));
+
+  if (!sni.empty()) {
+    EXPECT_CALL(socket_, setRequestedServerName(absl::string_view(sni)));
+  }
+
+  if (alpn.empty() || alpn[0] == 0) {
+    // No ALPN extension at all or empty ALPN list
+    EXPECT_CALL(socket_, setRequestedApplicationProtocols(_)).Times(0);
+  } else {
+    // Normal ALPN list
+    std::vector<absl::string_view> alpn_protos;
+    size_t start = 0;
+    while (start < alpn.size()) {
+      size_t len = static_cast<size_t>(alpn[start]);
+      alpn_protos.push_back(absl::string_view(alpn.data() + start + 1, len));
+      start += len + 1;
+    }
+    EXPECT_CALL(socket_, setRequestedApplicationProtocols(testing::ContainerEq(alpn_protos)));
+  }
+
+  EXPECT_CALL(socket_, setDetectedTransportProtocol(absl::string_view("tls")));
+  EXPECT_CALL(socket_, detectedTransportProtocol()).Times(::testing::AnyNumber());
+
+  EXPECT_TRUE(file_event_callback_(Event::FileReadyType::Read).ok());
+  auto state = filter_->onData(*buffer_);
+  EXPECT_EQ(Network::FilterStatus::Continue, state);
+}
+
+const absl::flat_hash_map<uint16_t, std::string> basic_test_version_to_ja4_ = {
```

Comment:
Where did these values come from? Are there published test vectors?

Reply:
No, I couldn't find any published test vectors. I tried a few cases where I used different browsers and looked at rthe computed hash on `browserleaks.com`. Other cases are mostly just comparing what the filter is computing so that we can make sure that if the hash changes in the future we can flag it.

Any recommendation on how to get the test vectors?

Reply:
If there are none published, one approach would be to use a few different clients to generate hashes from another known-good implementation, capture that handshake with tcpdump, and use those as our test vectors in envoy. You'd then change the test so that it doesn't use boringssl to initiate the client handshake, but just replay the client-hello you captured via tcpdump.

Reply:
Thanks, @ggreenway! I'll try to look at some already published test vectors and would do what you suggested. Any recommendations on how to programmatically replay the tcpdumps and write such tests?